### PR TITLE
Updated to correct service not obeying min instance

### PR DESCRIPTION
**Pull request by Anthony-McGreevy**
**JIRA**: https://portal.digitalsafe.net/browse/SCMOD-5400
**BUILD**: http://sou-jenkins2.swinfra.net/job/Autoscaler/view/Developer/job/Autoscaler~autoscaler~minInstanceFix~CI/

### CODE REVIEW

**Comment by michael-bryson** [comment]
JIRA: https://autjira.microfocus.com/browse/SCMOD-5400

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 1]
```diff
@@ -118,12 +118,10 @@ private void handleAnalysis()
...
-            if (firstRun) {
-                LOG.debug("Performing initial scaling checks for service {}", serviceRef);
-                action = handleFirstRun(instances);
-                firstRun = false;
-            } else {
-                action = analyser.analyseWorkload(instances);
+            LOG.debug("Performing scaling checks for service {}", serviceRef);
+            action = analyser.analyseWorkload(instances);
+            if(action.getOperation() == ScalingOperation.NONE){
+                action = handleServiceInstanceLimits(instances);
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: If we're no longer using `firstRun` then it should be removed.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java (hunk 1 above)
Comment (Anthony-McGreevy) [reply autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: variable removed

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
It seems like the governor has the responsibility for checking `maxInstances` when scaling operation is up so I don't see why it wouldn't also do it when scaling operation is none.

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 2]
```diff
@@ -118,23 +112,20 @@ private void handleAnalysis()
...
-                action = handleFirstRun(instances);
-                firstRun = false;
-            } else {
-                LOG.debug("Scaling checks for service {}", serviceRef);
-                action = analyser.analyseWorkload(instances);
-            }
-
+            LOG.debug("Performing scaling checks for service {}", serviceRef);
+            action = analyser.analyseWorkload(instances);
+            LOG.info("Workload Analyser determined that the autoscaler should {} {} by {} instances",
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: debug

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 3]
```diff
@@ -118,23 +112,20 @@ private void handleAnalysis()
...
+                     action.getOperation(), serviceRef, action.getAmount());
             action = governor.govern(serviceRef, action);
+            LOG.info("Governor determined that the autoscaler should {} {} by {} instances",
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: Debug if logging scale up and down elsewhere.

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java [hunk 4]
```diff
@@ -58,27 +58,60 @@ public ScalingAction govern(String serviceRef, ScalingAction action) {
...
+                        int target = Math.max(scalingConfiguration.getMinInstances(),
+                                              (int) Math.floor(lastInstanceInfo.getTotalInstances() * reduceToPercentage));
                         int amount = lastInstanceInfo.getTotalInstances() - target;
 
                         return new ScalingAction(ScalingOperation.SCALE_DOWN, amount);
                     }
+                } else {
+                    final int target = Math.min(scalingConfiguration.getMaxInstances() - lastInstanceInfo.getTotalInstances(),
+                                                Math.max(0, action.getAmount()));
+                    return new ScalingAction(action.getOperation(), target);
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java]: Just use `ScalingOperation.SCALE_UP`.

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java [hunk 5]
```diff
@@ -58,27 +58,60 @@ public ScalingAction govern(String serviceRef, ScalingAction action) {
...
 
-                        int target = Math.max(scalingConfiguration.getMinInstances(), (int)Math.floor(lastInstanceInfo.getTotalInstances() * reduceToPercentage));
+                        int target = Math.max(scalingConfiguration.getMinInstances(),
+                                              (int) Math.floor(lastInstanceInfo.getTotalInstances() * reduceToPercentage));
                         int amount = lastInstanceInfo.getTotalInstances() - target;
 
                         return new ScalingAction(ScalingOperation.SCALE_DOWN, amount);
                     }
+                } else {
+                    final int target = Math.min(scalingConfiguration.getMaxInstances() - lastInstanceInfo.getTotalInstances(),
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java]: `target` --> `delta`

**Review by dermot-hardy — APPROVED** [review]

**Pull request merged**